### ISSUE TITLE
preserve hyperlinks

### ISSUE THREAD

**Issue by januz** [qpdf/qpdf#94]
I'm searching for a command line solution to merge PDFs without losing hyperlinks within the source PDFs. After trying a bunch of different solutions, I was delighted to see that a pdf merged with qpdf preserved the links (i.e., when hovering over them with the mouse, one can click them), but sadly clicking the link doesn't result in anything. 

Is there a way to preserve the links? Thanks!

**Comment by bertalanimre** [comment]
As a user, I would say it won't happen. Simply because in my case, the links are targeting a specific page. If you merge more PDF files, then the pages are going to be re-counted and therefore the links will link you to the wrong page. At least this is what I think.

**Comment by jberkenbilt** [comment]
I don't think there's any deep reason why it wouldn't be doable, but it's complicated. I remember struggling with this issue when I added the capability of merging, and the documentation actually mentions this limitation though admittedly not in a way that's terribly clear: "All non-page data (info, outlines, page numbers, etc.) are taken from the primary input file."

It would be possible to create an enhancement to qpdf that would probably be able to handle this. It's been a long time since I've read that part of the spec, so I'm not sure how easy it would be. There could be issues such as clashing link targets that would require more work to resolve. Solvable, but not simple.

Let's leave this ticket open as a reminder. Unfortunately my time to work on qpdf is extremely limited right now, but I'm hoping this will eventually change and I'll be able to work on more enhancements, or maybe someone else will pick it up. If anyone is interested in trying to implement this, I would be willing to provide guidance.

**Comment by Nokia808** [comment]
@jberkenbilt 
Hi. This is a reminder about this issue because we, know, have PDF Mix Tool as a front-end for QPDF & need to fix this issue for better handling output file from original input to preserve metadata .....

**Comment by danielzgtg** [comment]
> As a user, I would say it won't happen. [...]

It is possible for a program to move the mouse, click on all the links, and note which page they lead to. It is probably not necessary to go that far and possible to analyze the PDF data instead. Therefore, qpdf can implement this feature.

**Comment by jberkenbilt** [comment]
I am not able to reproduce this issue, and there are now specific test cases for this, so I believe this was either subsequently fixed or was fixed in the extensive work I've done on fields and annotations in preparation for 10.2. I will be releasing 10.2 in the next few days. I'll comment here after it is released.

**Comment by jberkenbilt** [comment]
The version now in master contains all my annotation fixes. I will close this bug after releasing 10.2, but if it turns out the issue is still there, please feel free to reopen at that time.

**Comment by jberkenbilt** [comment]
Actually, I think I may be confused about what this is saying. Is it possible to provide a sample file? I know this issue is very old. I'm removing the next tag for now, but when I study document-level structures in more depth, I will likely get to this. This probably has to do with named destinations and probably only happens sometimes.

**Comment by AlisterH** [comment]
Could the problem be that people are using `--empty` to merge files easily, not realising that it throws this information away?
````
qpdf --empty --pages Contents.pdf empty_A4.pdf -- out.pdf
qpdf Contents.pdf --pages Contents.pdf 1-4 empty_A4.pdf -- out1.pdf
````
[Contents.pdf](https://github.com/qpdf/qpdf/files/6026944/Contents.pdf)
[empty_A4.pdf](https://github.com/qpdf/qpdf/files/6026945/empty_A4.pdf)
[out.pdf](https://github.com/qpdf/qpdf/files/6026946/out.pdf)
[out1.pdf](https://github.com/qpdf/qpdf/files/6026968/out1.pdf)

**Comment by jberkenbilt** [comment]
@AlisterH Yes, probably, but it shouldn't throw it away anyway. The current behavior is that all the document-level information is transferred unmodified from the main file. The better behavior would be that unused document-level information is discarded from the main file and relevant document-level information is copied from other files. 10.1 and earlier do this for page labels only. 10.2 will also do it for form fields. Going forward, I would like to do it for more and more document-level information. There will always be some differences between starting with `--empty` and starting with a file, like basic document /Info (producer, etc.), xmpp metadata, and some other stuff that is simply not possible to associate with content or pages, but as far as how the file behaves, any time when starting with `--empty` gives different results is probably undesirable. As more document-level stuff gets supported, I will introduce a mechanism to select which document-level stuff is carried forward. For example, it might be desirable in some cases to intentionally discard certain information. I don't know if that will come in 10.3 (or 11.0 or whatever follows 10.2) or later, but it's on my roadmap.

In any case, you make a good point that, for now, starting with the file whose behavior you want to maintain gives better results than starting with `--empty`.

**Comment by sskras** [comment]
@jberkenbilt commented [on Feb 23, 2021](#issuecomment-784135026):

> @AlisterH Yes, probably, but it shouldn't throw it away anyway. [...] 
> 
> In any case, you make a good point that, for now, starting with the file whose behavior you want to maintain gives better results than starting with `--empty`.

Thank you both for the hint!

> as far as how the file behaves, any time when starting with `--empty` gives different results is probably undesirable. As more document-level stuff gets supported, I will introduce a mechanism to select which document-level stuff is carried forward. For example, it might be desirable in some cases to intentionally discard certain information. I don't know if that will come in 10.3 (or 11.0 or whatever follows 10.2) or later, but it's on my roadmap.

Thanks once more for the update.
I have checked this using v10.4:
```
$ /mingw64/bin/qpdf --version
qpdf version 10.4.0
Run qpdf --copyright to see copyright and license information.
```

... and it still destroys my TOC from the second document in a row when using `--empty`.

**Comment by jberkenbilt** [comment]
This issue is part of the qpdf pages epic. If you are interested in following, please see #1104.
